forem/forem
# Create DiscussionLocks

## What type of PR is this? (check all applicable)

## Description
This PR introduces functionality to allow **article authors, admins, and super admins** to lock discussions which basically prevents the ability for users to add any further comments to an article. We've also added some logic to allow a permitted user to include a `reason` that will display publicly when a discussion is locked and `notes` which are only visible to users permitted to lock the discussion.

## Related Tickets & Documents
closes https://github.com/forem/forem/issues/9619

## QA Instructions, Screenshots, Recordings
1. Pull down this code.
2. Fire up the Rails server locally - `bin/startup`.
3. Login.
4. Write a post.
5. Go to `/manage` on your published post.
6. Click `Lock discussion`.
7. Add a `reason` and `notes`.
8. Lock the discussion.
9. View your article - you should see the discussion is locked and you're unable to add a comment.
10. Go back to `/manage` on your post.
11. Unlock the discussion.
12. Go back to your post and leave a comment.

**Lock discussion**
![lock_discussion](https://user-images.githubusercontent.com/15987080/120568772-91bc6400-c3e2-11eb-8269-259be92d3455.png)

**Confirm lock discussion**
![lock_discussion_confirm](https://user-images.githubusercontent.com/15987080/120568967-fe376300-c3e2-11eb-8981-dc0109a65107.png)

**Lock discussion success**
![success](https://user-images.githubusercontent.com/15987080/120569018-19a26e00-c3e3-11eb-99c2-1c1983d07695.png)

**Post with discussion locked (with reason)**
![with_reason](https://user-images.githubusercontent.com/15987080/120569048-27f08a00-c3e3-11eb-84fa-aaab56d8e8fe.png)

**Post with discussion locked (without reason)**
![without_reason](https://user-images.githubusercontent.com/15987080/120569081-3343b580-c3e3-11eb-98b9-0a861edb014b.png)

**Unlock discussion**
![unlock_discussion](https://user-images.githubusercontent.com/15987080/120569116-422a6800-c3e3-11eb-9d10-f2e232288bab.png)

**Confirm unlock discussion**
![unlock_discussion_confirm](https://user-images.githubusercontent.com/15987080/120569202-71d97000-c3e3-11eb-8408-a5b905d79c59.png)

**Error example**
![error](https://user-images.githubusercontent.com/15987080/120569217-7736ba80-c3e3-11eb-9867-bc559684c8c6.png)

## Added tests?

## [Forem core team only] How will this change be communicated?

![wedding_crashers_lock_it_up_gif](https://media.giphy.com/media/3ohzdFvSAgvmXq6n2o/giphy.gif)
